File: backend/app/core/sequence_risk.py
"""
Sequence risk stress testing engine.
Applies historical crisis return sequences to the first N years of retirement,
then runs Monte Carlo for the remainder.
"""

import numpy as np

from app.core.monte_carlo import (
    _build_covariance_matrix,
    _compute_withdrawals_for_year,
    _generate_returns_parametric,
)
from app.core.withdrawal_strategies import resolve_initial_rate


def _run_single_scenario(
    rng: np.random.Generator,
    n_sims: int,
    initial_portfolio: float,
    weights: np.ndarray,
    expected_returns: np.ndarray,
    cov_matrix: np.ndarray,
    n_years_decum: int,
    retirement_age: int,
    strategy: str,
    strategy_params: dict,
    inflation: float,
    expense_ratio: float,
    post_retirement_income: list,
    crisis_returns: np.ndarray | None = None,
) -> dict:
    """Run MC simulation with optional crisis return sequence in early years."""
    n_crisis = len(crisis_returns) if crisis_returns is not None else 0

    # Generate parametric returns for all years
    all_returns = _generate_returns_parametric(
        rng, n_sims, n_years_decum, weights, expected_returns, cov_matrix
    )

    # Override early years with crisis returns (applied to all sims)
    if crisis_returns is not None and n_crisis > 0:
        apply_years = min(n_crisis, n_years_decum)
        for y in range(apply_years):
            all_returns[:, y] = crisis_returns[y]

    # Simulate
    balances = np.zeros((n_sims, n_years_decum + 1))
    balances[:, 0] = initial_portfolio
    failed = np.zeros(n_sims, dtype=bool)
    failure_year = np.full(n_sims, n_years_decum)

    prev_withdrawal = np.zeros(n_sims)
    swr = resolve_initial_rate(strategy_params)
    initial_withdrawal_amount = initial_portfolio * swr

    for t in range(n_years_decum):
        # Pass previous year's return for Guyton-Klinger PMR
        pyr = all_returns[:, t - 1] if t > 0 else None

        withdrawal = _compute_withdrawals_for_year(
            strategy=strategy,
            portfolio=balances[:, t],
            year=t,
            n_years_decum=n_years_decum,
            initial_withdrawal=initial_withdrawal_amount,
            prev_withdrawals=prev_withdrawal,
            inflation=inflation,
            strategy_params=strategy_params,
            prev_year_return=pyr,
        )

        income = post_retirement_income[t] if t < len(post_retirement_income) else 0.0
        net_withdrawal = np.maximum(0, withdrawal - income)
        net_withdrawal = np.minimum(net_withdrawal, balances[:, t])

        prev_withdrawal = withdrawal

        balances[:, t + 1] = (
            (balances[:, t] - net_withdrawal) * (1 + all_returns[:, t] - expense_ratio)
        )

        newly_failed = (balances[:, t + 1] <= 0) & (~failed)
        failed[newly_failed] = True
        failure_year[newly_failed] = t
        balances[:, t + 1] = np.maximum(balances[:, t + 1], 0)

    # Compute percentile bands
    percentile_values = [5, 10, 25, 50, 75, 90, 95]
    percentiles = np.percentile(balances, percentile_values, axis=0)

    years = list(range(n_years_decum + 1))
    ages = [retirement_age + y for y in years]

    percentile_bands = {
        "years": years,
        "ages": ages,
        "p5": percentiles[0].tolist(),
        "p10": percentiles[1].tolist(),
        "p25": percentiles[2].tolist(),
        "p50": percentiles[3].tolist(),
        "p75": percentiles[4].tolist(),
        "p90": percentiles[5].tolist(),
        "p95": percentiles[6].tolist(),
    }

    return {
        "success_rate": float(1 - np.mean(failed)),
        "percentile_bands": percentile_bands,
    }


def _run_mitigation(
    rng: np.random.Generator,
    n_sims: int,
    initial_portfolio: float,
    weights: np.ndarray,
    expected_returns: np.ndarray,
    cov_matrix: np.ndarray,
    n_years_decum: int,
    retirement_age: int,
    strategy: str,
    strategy_params: dict,
    inflation: float,
    expense_ratio: float,
    post_retirement_income: list,
    crisis_returns: np.ndarray | None,
    mitigation_name: str,
    mitigation_desc: str,
    baseline_crisis_rate: float,
    modified_weights: np.ndarray | None = None,
    modified_strategy_params: dict | None = None,
) -> dict:
    """Run a single mitigation scenario and return impact vs unmitigated baseline."""
    eff_weights = modified_weights if modified_weights is not None else weights
    eff_params = modified_strategy_params if modified_strategy_params is not None else strategy_params

    normal = _run_single_scenario(
        rng=np.random.default_rng(rng.integers(0, 2**32)),
        n_sims=n_sims,
        initial_portfolio=initial_portfolio,
        weights=eff_weights,
        expected_returns=expected_returns,
        cov_matrix=cov_matrix,
        n_years_decum=n_years_decum,
        retirement_age=retirement_age,
        strategy=strategy,
        strategy_params=eff_params,
        inflation=inflation,
        expense_ratio=expense_ratio,
        post_retirement_income=post_retirement_income,
        crisis_returns=None,
    )

    crisis = _run_single_scenario(
        rng=np.random.default_rng(rng.integers(0, 2**32)),
        n_sims=n_sims,
        initial_portfolio=initial_portfolio,
        weights=eff_weights,
        expected_returns=expected_returns,
        cov_matrix=cov_matrix,
        n_years_decum=n_years_decum,
        retirement_age=retirement_age,
        strategy=strategy,
        strategy_params=eff_params,
        inflation=inflation,
        expense_ratio=expense_ratio,
        post_retirement_income=post_retirement_income,
        crisis_returns=crisis_returns,
    )

    return {
        "strategy": mitigation_name,
        "description": mitigation_desc,
        "normal_success_rate": normal["success_rate"],
        "crisis_success_rate": crisis["success_rate"],
        "success_improvement": crisis["success_rate"] - baseline_crisis_rate,
    }


def run_sequence_risk(params: dict) -> dict:
    """
    Run sequence risk analysis comparing normal vs crisis scenarios.

    Includes 3 mitigation strategies:
    - Bond tent: increase bond allocation in early retirement
    - Cash buffer: hold 2 years expenses in cash
    - Flexible spending: reduce withdrawal rate by 15%
    """
    n_sims = params.get("n_simulations", 2000)
    seed = params.get("seed")
    rng = np.random.default_rng(seed)

    initial_portfolio = params["initial_portfolio"]
    weights = np.array(params["allocation_weights"])
    expected_returns = np.array(params["expected_returns"])
    std_devs = np.array(params["std_devs"])
    correlation_matrix = np.array(params["correlation_matrix"])
    expense_ratio = params.get("expense_ratio", 0.003)
    inflation = params.get("inflation", 0.025)

    retirement_age = params["retirement_age"]
    life_expectancy = params["life_expectancy"]
    n_years_decum = max(1, life_expectancy - retirement_age)

    strategy = params.get("withdrawal_strategy", "constant_dollar")
    strategy_params = params.get("strategy_params", {})
    post_retirement_income = params.get("post_retirement_income", [])

    crisis = params.get("crisis", {})
    crisis_returns = np.array(crisis.get("equity_return_sequence", []))

    cov_matrix = _build_covariance_matrix(std_devs, correlation_matrix)

    # Normal scenario
    normal_result = _run_single_scenario(
        rng=np.random.default_rng(rng.integers(0, 2**32)),
        n_sims=n_sims,
        initial_portfolio=initial_portfolio,
        weights=weights,
        expected_returns=expected_returns,
        cov_matrix=cov_matrix,
        n_years_decum=n_years_decum,
        retirement_age=retirement_age,
        strategy=strategy,
        strategy_params=strategy_params,
        inflation=inflation,
        expense_ratio=expense_ratio,
        post_retirement_income=post_retirement_income,
        crisis_returns=None,
    )

    # Crisis scenario
    crisis_result = _run_single_scenario(
        rng=np.random.default_rng(rng.integers(0, 2**32)),
        n_sims=n_sims,
        initial_portfolio=initial_portfolio,
        weights=weights,
        expected_returns=expected_returns,
        cov_matrix=cov_matrix,
        n_years_decum=n_years_decum,
        retirement_age=retirement_age,
        strategy=strategy,
        strategy_params=strategy_params,
        inflation=inflation,
        expense_ratio=expense_ratio,
        post_retirement_income=post_retirement_income,
        crisis_returns=crisis_returns,
    )

    # Mitigations
    mitigations = []

    # 1. Bond tent: shift 20% from equities to bonds in early retirement
    bond_tent_weights = weights.copy()
    # Move 20% from equity classes (0,1,2) to bonds (3)
    equity_total = bond_tent_weights[0] + bond_tent_weights[1] + bond_tent_weights[2]
    shift = min(0.20, equity_total)
    if equity_total > 0:
        for i in [0, 1, 2]:
            reduction = shift * (bond_tent_weights[i] / equity_total)
            bond_tent_weights[i] -= reduction
        bond_tent_weights[3] += shift

    baseline_crisis_rate = crisis_result["success_rate"]

    mitigations.append(_run_mitigation(
        rng=np.random.default_rng(rng.integers(0, 2**32)),
        n_sims=n_sims,
        initial_portfolio=initial_portfolio,
        weights=bond_tent_weights,
        expected_returns=expected_returns,
        cov_matrix=cov_matrix,
        n_years_decum=n_years_decum,
        retirement_age=retirement_age,
        strategy=strategy,
        strategy_params=strategy_params,
        inflation=inflation,
        expense_ratio=expense_ratio,
        post_retirement_income=post_retirement_income,
        crisis_returns=crisis_returns,
        mitigation_name="Conservative Allocation",
        mitigation_desc="Shift 20% from equities to bonds throughout retirement to reduce volatility exposure.",
        baseline_crisis_rate=baseline_crisis_rate,
    ))

    # 2. Cash buffer: hold 2 years expenses in cash (reduce portfolio, zero-risk buffer)
    annual_expense_est = initial_portfolio * resolve_initial_rate(strategy_params)
    cash_buffer = annual_expense_est * 2
    reduced_portfolio = max(initial_portfolio - cash_buffer, initial_portfolio * 0.5)

    mitigations.append(_run_mitigation(
        rng=np.random.default_rng(rng.integers(0, 2**32)),
        n_sims=n_sims,
        initial_portfolio=reduced_portfolio,
        weights=weights,
        expected_returns=expected_returns,
        cov_matrix=cov_matrix,
        n_years_decum=n_years_decum,
        retirement_age=retirement_age,
        strategy=strategy,
        strategy_params=strategy_params,
        inflation=inflation,
        expense_ratio=expense_ratio,
        post_retirement_income=[inc + cash_buffer / min(2, n_years_decum) for inc in (post_retirement_income[:2] if len(post_retirement_income) >= 2 else [0.0, 0.0])] + post_retirement_income[2:],
        crisis_returns=crisis_returns,
        mitigation_name="Cash Buffer (2 Years)",
        mitigation_desc="Hold 2 years of expenses in cash outside the portfolio, drawing from buffer in early crisis years.",
        baseline_crisis_rate=baseline_crisis_rate,
    ))

    # 3. Flexible spending: reduce SWR by 15%
    flexible_params = dict(strategy_params)
    if "swr" in flexible_params:
        flexible_params["swr"] = flexible_params["swr"] * 0.85
    if "initial_rate" in flexible_params:
        flexible_params["initial_rate"] = flexible_params["initial_rate"] * 0.85
    if "target_rate" in flexible_params:
        flexible_params["target_rate"] = flexible_params["target_rate"] * 0.85

    mitigations.append(_run_mitigation(
        rng=np.random.default_rng(rng.integers(0, 2**32)),
        n_sims=n_sims,
        initial_portfolio=initial_portfolio,
        weights=weights,
        expected_returns=expected_returns,
        cov_matrix=cov_matrix,
        n_years_decum=n_years_decum,
        retirement_age=retirement_age,
        strategy=strategy,
        strategy_params=flexible_params,
        inflation=inflation,
        expense_ratio=expense_ratio,
        post_retirement_income=post_retirement_income,
        crisis_returns=crisis_returns,
        mitigation_name="Flexible Spending (-15%)",
        mitigation_desc="Reduce withdrawal rate by 15% to preserve capital during market downturns.",
        baseline_crisis_rate=baseline_crisis_rate,
        modified_strategy_params=flexible_params,
    ))

    return {
        "normal_success_rate": normal_result["success_rate"],
        "crisis_success_rate": crisis_result["success_rate"],
        "success_degradation": normal_result["success_rate"] - crisis_result["success_rate"],
        "normal_percentile_bands": normal_result["percentile_bands"],
        "crisis_percentile_bands": crisis_result["percentile_bands"],
        "mitigations": mitigations,
    }

File: frontend/src/lib/simulation/sequenceRisk.ts
/**
 * Sequence risk stress testing engine.
 * TypeScript port of backend/app/core/sequence_risk.py (349 lines).
 *
 * Applies historical crisis return sequences to the first N years of retirement,
 * then runs Monte Carlo for the remainder.
 *
 * Reuses MC internals from monteCarlo.ts:
 *   - generateReturnsParametric
 *   - computeWithdrawalsForYear
 *   - resolveInitialRate
 *
 * Formulas from FIRE_PLANNER_MASTER_PLAN_v2.md Section 6.
 */

import { SeededRNG } from '@/lib/math/random.ts'
import { choleskyDecomposition, buildCovarianceMatrix } from '@/lib/math/linalg.ts'
import { percentile } from '@/lib/math/stats.ts'
import {
  generateReturnsParametric,
  computeWithdrawalsForYear,
  resolveInitialRate,
} from './monteCarlo.ts'
import type { PercentileBands, MitigationImpact, RetirementMitigationConfig } from '@/lib/types.ts'

// ============================================================
// Types
// ============================================================

export interface SequenceRiskEngineParams {
  initialPortfolio: number
  allocationWeights: number[]
  expectedReturns: number[]
  stdDevs: number[]
  correlationMatrix: number[][]
  retirementAge: number
  lifeExpectancy: number
  withdrawalStrategy: string
  strategyParams: Record<string, number>
  nSimulations: number
  seed?: number
  expenseRatio: number
  inflation: number
  postRetirementIncome: number[]
  oneTimeWithdrawals?: { year: number; amount: number }[]
  portfolioInjections?: { year: number; amount: number }[]
  retirementMitigation?: RetirementMitigationConfig
  annualExpensesAtRetirement?: number
  crisis: {
    id: string
    name: string
    equityReturnSequence: number[]
    durationYears: number
  }
}

export interface SequenceRiskEngineResult {
  normal_success_rate: number
  crisis_success_rate: number
  success_degradation: number
  normal_percentile_bands: PercentileBands
  crisis_percentile_bands: PercentileBands
  mitigations: MitigationImpact[]
}

interface SingleScenarioResult {
  success_rate: number
  percentile_bands: PercentileBands
}

// ============================================================
// Single Scenario Runner (maps to _run_single_scenario in Python)
// ============================================================

/**
 * Run a simplified MC simulation (decumulation only) with optional crisis
 * return overrides for the first N years.
 *
 * @param rng - Child RNG for this scenario
 * @param nSims - Number of simulation paths
 * @param initialPortfolio - Starting portfolio value
 * @param weights - 8-element allocation weights
 * @param expectedReturns - 8-element expected annual returns
 * @param stdDevs - 8-element standard deviations
 * @param correlationMatrix - 8×8 correlation matrix
 * @param precomputedL - Pre-computed Cholesky factor (pass to avoid recomputation)
 * @param nYearsDecum - Number of decumulation years
 * @param retirementAge - Age at retirement (for ages array in bands)
 * @param strategy - Withdrawal strategy name
 * @param strategyParams - Strategy-specific parameters
 * @param inflation - Annual inflation rate
 * @param expenseRatio - Annual portfolio expense ratio
 * @param postRetirementIncome - Per-year supplemental income array
 * @param crisisReturns - Optional sequence of portfolio returns overriding early years
 */
function runSingleScenario(
  rng: SeededRNG,
  nSims: number,
  initialPortfolio: number,
  weights: number[],
  expectedReturns: number[],
  stdDevs: number[],
  correlationMatrix: number[][],
  precomputedL: number[][],
  nYearsDecum: number,
  retirementAge: number,
  strategy: string,
  strategyParams: Record<string, number>,
  inflation: number,
  expenseRatio: number,
  postRetirementIncome: number[],
  crisisReturns: number[] | null,
  oneTimeWithdrawals?: { year: number; amount: number }[],
  annualExpensesAtRetirement?: number,
  portfolioInjections?: { year: number; amount: number }[],
): SingleScenarioResult {
  const nCrisis = crisisReturns !== null ? crisisReturns.length : 0

  // Generate parametric returns for all years: [nSims][nYearsDecum]
  const allReturns = generateReturnsParametric(
    rng,
    nSims,
    nYearsDecum,
    weights,
    expectedReturns,
    stdDevs,
    correlationMatrix,
    precomputedL,
  )

  // Override early years with crisis returns (applied uniformly to all sims)
  if (crisisReturns !== null && nCrisis > 0) {
    const applyYears = Math.min(nCrisis, nYearsDecum)
    for (let y = 0; y < applyYears; y++) {
      for (let s = 0; s < nSims; s++) {
        allReturns[s][y] = crisisReturns[y]
      }
    }
  }

  // Simulation state
  const balances: number[][] = Array.from({ length: nSims }, () => {
    const arr = new Array(nYearsDecum + 1).fill(0)
    arr[0] = initialPortfolio
    return arr
  })

  const failed: boolean[] = new Array(nSims).fill(false)
  const failureYear: number[] = new Array(nSims).fill(nYearsDecum)
  const prevWithdrawals: number[] = new Array(nSims).fill(0)
  // Track previous balance for sensible_withdrawals gain calculation
  const prevBalances: number[] = new Array(nSims).fill(initialPortfolio)

  const swr = resolveInitialRate(strategyParams)
  const expenses = annualExpensesAtRetirement ?? 0
  // Use user's actual retirement expenses when available;
  // fall back to portfolio × SWR rate when no expenses are specified.
  const initialWithdrawalAmount = expenses > 0
    ? expenses
    : initialPortfolio * swr

  for (let t = 0; t < nYearsDecum; t++) {
    // Apply portfolio injections (e.g. downsizing equity) at this year
    if (portfolioInjections) {
      for (const inj of portfolioInjections) {
        if (inj.year === t) {
          for (let s = 0; s < nSims; s++) {
            balances[s][t] += inj.amount
          }
        }
      }
    }

    for (let s = 0; s < nSims; s++) {
      const prevYearReturn = t > 0 ? allReturns[s][t - 1] : undefined

      // Previous year gains for sensible_withdrawals
      const prevYearGains = t > 0
        ? balances[s][t] - prevBalances[s] + prevWithdrawals[s]
        : 0

      const withdrawal = computeWithdrawalsForYear(
        strategy,
        balances[s][t],
        t,
        nYearsDecum,
        initialWithdrawalAmount,
        prevWithdrawals[s],
        inflation,
        strategyParams,
        prevYearReturn,
        prevYearGains,
      )

      // Add one-time withdrawals for this year offset
      const oneTime = (oneTimeWithdrawals ?? [])
        .filter(w => w.year === t)
        .reduce((sum, w) => sum + w.amount, 0)

      const income = t < postRetirementIncome.length ? postRetirementIncome[t] : 0.0
      let netWithdrawal = Math.max(0, (withdrawal + oneTime) - income)
      // Don't withdraw more than the portfolio
      netWithdrawal = Math.min(netWithdrawal, balances[s][t])

      prevWithdrawals[s] = withdrawal
      prevBalances[s] = balances[s][t]

      balances[s][t + 1] =
        (balances[s][t] - netWithdrawal) * (1 + allReturns[s][t] - expenseRatio)

      if (balances[s][t + 1] <= 0 && !failed[s]) {
        failed[s] = true
        failureYear[s] = t
      }
      // Clamp to 0
      if (balances[s][t + 1] < 0) {
        balances[s][t + 1] = 0
      }
    }
  }

  // Compute percentile bands across simulations at each time point
  const nPoints = nYearsDecum + 1
  const years: number[] = []
  const ages: number[] = []
  const p5: number[] = []
  const p10: number[] = []
  const p25: number[] = []
  const p50: number[] = []
  const p75: number[] = []
  const p90: number[] = []
  const p95: number[] = []

  for (let y = 0; y < nPoints; y++) {
    years.push(y)
    ages.push(retirementAge + y)

    const col: number[] = new Array(nSims)
    for (let s = 0; s < nSims; s++) {
      col[s] = balances[s][y]
    }

    p5.push(percentile(col, 5))
    p10.push(percentile(col, 10))
    p25.push(percentile(col, 25))
    p50.push(percentile(col, 50))
    p75.push(percentile(col, 75))
    p90.push(percentile(col, 90))
    p95.push(percentile(col, 95))
  }

  const nFailed = failed.filter(Boolean).length
  const successRate = 1 - nFailed / nSims

  return {
    success_rate: successRate,
    percentile_bands: { years, ages, p5, p10, p25, p50, p75, p90, p95 },
  }
}

// ============================================================
// Mitigation Runner (maps to _run_mitigation in Python)
// ============================================================

interface RunMitigationParams {
  rng: SeededRNG
  nSims: number
  initialPortfolio: number
  weights: number[]
  expectedReturns: number[]
  stdDevs: number[]
  correlationMatrix: number[][]
  precomputedL: number[][]
  nYearsDecum: number
  retirementAge: number
  strategy: string
  strategyParams: Record<string, number>
  inflation: number
  expenseRatio: number
  postRetirementIncome: number[]
  crisisReturns: number[]
  oneTimeWithdrawals?: { year: number; amount: number }[]
  portfolioInjections?: { year: number; amount: number }[]
  annualExpensesAtRetirement?: number
  mitigationName: string
  mitigationDesc: string
  baselineCrisisRate: number
  modifiedWeights?: number[]
  modifiedStrategyParams?: Record<string, number>
  modifiedPostRetirementIncome?: number[]
  modifiedInitialPortfolio?: number
}

function runMitigation(p: RunMitigationParams): MitigationImpact {
  const effWeights = p.modifiedWeights ?? p.weights
  const effParams = p.modifiedStrategyParams ?? p.strategyParams
  const effPortfolio = p.modifiedInitialPortfolio ?? p.initialPortfolio
  const effIncome = p.modifiedPostRetirementIncome ?? p.postRetirementIncome

  // Normal run (no crisis) for this mitigation
  const normalResult = runSingleScenario(
    new SeededRNG(p.rng.nextInt(2 ** 32)),
    p.nSims,
    effPortfolio,
    effWeights,
    p.expectedReturns,
    p.stdDevs,
    p.correlationMatrix,
    p.precomputedL,
    p.nYearsDecum,
    p.retirementAge,
    p.strategy,
    effParams,
    p.inflation,
    p.expenseRatio,
    effIncome,
    null,
    p.oneTimeWithdrawals,
    p.annualExpensesAtRetirement,
    p.portfolioInjections,
  )

  // Crisis run for this mitigation
  const crisisResult = runSingleScenario(
    new SeededRNG(p.rng.nextInt(2 ** 32)),
    p.nSims,
    effPortfolio,
    effWeights,
    p.expectedReturns,
    p.stdDevs,
    p.correlationMatrix,
    p.precomputedL,
    p.nYearsDecum,
    p.retirementAge,
    p.strategy,
    effParams,
    p.inflation,
    p.expenseRatio,
    effIncome,
    p.crisisReturns,
    p.oneTimeWithdrawals,
    p.annualExpensesAtRetirement,
    p.portfolioInjections,
  )

  return {
    strategy: p.mitigationName,
    description: p.mitigationDesc,
    normal_success_rate: normalResult.success_rate,
    crisis_success_rate: crisisResult.success_rate,
    success_improvement: crisisResult.success_rate - p.baselineCrisisRate,
  }
}

// ============================================================
// Main Entry Point
// ============================================================

/**
 * Run sequence risk stress testing.
 *
 * Applies a historical crisis return sequence to the first N years of retirement,
 * then runs parametric Monte Carlo for the remainder. Returns normal vs crisis
 * success rates, percentile bands, and 3 mitigation strategy comparisons.
 *
 * Output uses snake_case keys matching SequenceRiskResult from types.ts.
 */
export function runSequenceRisk(params: SequenceRiskEngineParams): SequenceRiskEngineResult {
  const {
    initialPortfolio,
    allocationWeights: weights,
    expectedReturns,
    stdDevs,
    correlationMatrix,
    retirementAge,
    lifeExpectancy,
    withdrawalStrategy: strategy,
    strategyParams,
    nSimulations: nSims,
    seed,
    expenseRatio,
    inflation,
    postRetirementIncome,
    oneTimeWithdrawals,
    portfolioInjections,
    annualExpensesAtRetirement,
    crisis,
  } = params

  const masterRng = new SeededRNG(seed ?? 42)
  const nYearsDecum = Math.max(1, lifeExpectancy - retirementAge)
  const crisisReturns = crisis.equityReturnSequence

  // Pre-compute Cholesky factor once — reused across all scenario runs
  const covMatrix = buildCovarianceMatrix(stdDevs, correlationMatrix)
  const L = choleskyDecomposition(covMatrix)

  // --- Normal scenario (no crisis) ---
  const normalResult = runSingleScenario(
    new SeededRNG(masterRng.nextInt(2 ** 32)),
    nSims,
    initialPortfolio,
    weights,
    expectedReturns,
    stdDevs,
    correlationMatrix,
    L,
    nYearsDecum,
    retirementAge,
    strategy,
    strategyParams,
    inflation,
    expenseRatio,
    postRetirementIncome,
    null,
    oneTimeWithdrawals,
    annualExpensesAtRetirement,
    portfolioInjections,
  )

  // --- Crisis scenario ---
  const crisisResult = runSingleScenario(
    new SeededRNG(masterRng.nextInt(2 ** 32)),
    nSims,
    initialPortfolio,
    weights,
    expectedReturns,
    stdDevs,
    correlationMatrix,
    L,
    nYearsDecum,
    retirementAge,
    strategy,
    strategyParams,
    inflation,
    expenseRatio,
    postRetirementIncome,
    crisisReturns.length > 0 ? crisisReturns : null,
    oneTimeWithdrawals,
    annualExpensesAtRetirement,
    portfolioInjections,
  )

  const baselineCrisisRate = crisisResult.success_rate

  // ============================================================
  // Mitigation 1: Conservative Allocation (Bond Tent)
  // Shift 20% from equities (indices 0,1,2) to bonds (index 3),
  // proportional to each equity's current weight.
  // ============================================================
  const bondTentWeights = [...weights]
  const equityTotal = bondTentWeights[0] + bondTentWeights[1] + bondTentWeights[2]
  const shift = Math.min(0.20, equityTotal)
  if (equityTotal > 0) {
    for (const i of [0, 1, 2] as const) {
      const reduction = shift * (bondTentWeights[i] / equityTotal)
      bondTentWeights[i] -= reduction
    }
    bondTentWeights[3] += shift
  }

  const mitigationConservativeAlloc = runMitigation({
    rng: new SeededRNG(masterRng.nextInt(2 ** 32)),
    nSims,
    initialPortfolio,
    weights,
    expectedReturns,
    stdDevs,
    correlationMatrix,
    precomputedL: L,
    nYearsDecum,
    retirementAge,
    strategy,
    strategyParams,
    inflation,
    expenseRatio,
    postRetirementIncome,
    crisisReturns,
    oneTimeWithdrawals,
    portfolioInjections,
    annualExpensesAtRetirement,
    mitigationName: 'Conservative Allocation',
    mitigationDesc:
      'Shift 20% from equities to bonds throughout retirement to reduce volatility exposure.',
    baselineCrisisRate,
    modifiedWeights: bondTentWeights,
  })

  // ============================================================
  // Mitigation 2: Cash Buffer (2 Years)
  // Hold 2 years of expenses outside the invested portfolio.
  // Reduce invested portfolio by that amount (floored at 50% of original),
  // and add back cash buffer spread over the first 2 decumulation years.
  // ============================================================
  const expensesAtRetirement = annualExpensesAtRetirement ?? 0
  const annualExpenseEst = expensesAtRetirement > 0
    ? expensesAtRetirement
    : initialPortfolio * resolveInitialRate(strategyParams)
  const cashBuffer = annualExpenseEst * 2
  const reducedPortfolio = Math.max(
    initialPortfolio - cashBuffer,
    initialPortfolio * 0.5,
  )

  // Add buffer equally across the first min(2, nYearsDecum) years
  const bufferYears = Math.min(2, nYearsDecum)
  const cashBufferPerYear = cashBuffer / bufferYears

  // Build modified post-retirement income: add cash buffer to years 0 and 1
  const modifiedIncome: number[] = []
  for (let i = 0; i < nYearsDecum; i++) {
    const base = i < postRetirementIncome.length ? postRetirementIncome[i] : 0
    const buffer = i < bufferYears ? cashBufferPerYear : 0
    modifiedIncome.push(base + buffer)
  }

  const mitigationCashBuffer = runMitigation({
    rng: new SeededRNG(masterRng.nextInt(2 ** 32)),
    nSims,
    initialPortfolio,
    weights,
    expectedReturns,
    stdDevs,
    correlationMatrix,
    precomputedL: L,
    nYearsDecum,
    retirementAge,
    strategy,
    strategyParams,
    inflation,
    expenseRatio,
    postRetirementIncome,
    crisisReturns,
    oneTimeWithdrawals,
    portfolioInjections,
    annualExpensesAtRetirement,
    mitigationName: 'Cash Buffer (2 Years)',
    mitigationDesc:
      'Hold 2 years of expenses in cash outside the portfolio, drawing from buffer in early crisis years.',
    baselineCrisisRate,
    modifiedInitialPortfolio: reducedPortfolio,
    modifiedPostRetirementIncome: modifiedIncome,
  })

  // ============================================================
  // Mitigation 3: Flexible Spending (-15%)
  // Reduce SWR / initial rate / target rate by 15%.
  // Uses camelCase keys (TS convention) with fallback to snake_case.
  // ============================================================
  const flexibleParams: Record<string, number> = { ...strategyParams }
  if ('swr' in flexibleParams) flexibleParams.swr *= 0.85
  if ('initialRate' in flexibleParams) flexibleParams.initialRate *= 0.85
  if ('initial_rate' in flexibleParams) flexibleParams.initial_rate *= 0.85
  if ('targetRate' in flexibleParams) flexibleParams.targetRate *= 0.85
  if ('target_rate' in flexibleParams) flexibleParams.target_rate *= 0.85

  const mitigationFlexibleSpending = runMitigation({
    rng: new SeededRNG(masterRng.nextInt(2 ** 32)),
    nSims,
    initialPortfolio,
    weights,
    expectedReturns,
    stdDevs,
    correlationMatrix,
    precomputedL: L,
    nYearsDecum,
    retirementAge,
    strategy,
    strategyParams,
    inflation,
    expenseRatio,
    postRetirementIncome,
    crisisReturns,
    oneTimeWithdrawals,
    portfolioInjections,
    annualExpensesAtRetirement: annualExpensesAtRetirement
      ? annualExpensesAtRetirement * 0.85
      : undefined,
    mitigationName: 'Flexible Spending (-15%)',
    mitigationDesc:
      'Reduce withdrawal rate by 15% to preserve capital during market downturns.',
    baselineCrisisRate,
    modifiedStrategyParams: flexibleParams,
  })

  return {
    normal_success_rate: normalResult.success_rate,
    crisis_success_rate: crisisResult.success_rate,
    success_degradation: normalResult.success_rate - crisisResult.success_rate,
    normal_percentile_bands: normalResult.percentile_bands,
    crisis_percentile_bands: crisisResult.percentile_bands,
    mitigations: [
      mitigationConservativeAlloc,
      mitigationCashBuffer,
      mitigationFlexibleSpending,
    ],
  }
}
